wpilibj/src/main/java/edu/wpi/first/wpilibj/AllianceFlipper.java

```diff
@@ -0,0 +1,254 @@
+ *   }
+ * }
+ * </code></pre>
+ */
+public class AllianceFlipper {
+  private AllianceFlipper() {
+    throw new UnsupportedOperationException("This is a utility class!");
+  }
+
+  /** The flipper to use for flipping coordinates. */
+  public static enum Flipper {
+    /**
+     * X becomes fieldLength - x, leaves the y coordinate unchanged, and heading becomes PI -
+     * heading.
+     */
+    VERTICALLY_MIRRORED {
+      public double flipX(double x) {
+        return activeYear.fieldLength - x;
+      }
+
+      public double flipY(double y) {
+        return y;
+      }
+
+      public double flipHeading(double heading) {
+        return Math.PI - heading;
+      }
+    },
+    /** X becomes fieldLength - x, Y becomes fieldWidth - y, and heading becomes PI - heading. */
+    ROTATIONALLY_MIRRORED {
+      public double flipX(double x) {
+        return activeYear.fieldLength - x;
+      }
+
+      public double flipY(double y) {
+        return activeYear.fieldWidth - y;
+      }
+
+      public double flipHeading(double heading) {
+        return Math.PI - heading;
+      }
+    };
+
+    /**
+     * Flips the X coordinate.
+     *
+     * @param x The X coordinate to flip.
+     * @return The flipped X coordinate.
+     */
+    public abstract double flipX(double x);
+
+    /**
+     * Flips the Y coordinate.
+     *
+     * @param y The Y coordinate to flip.
+     * @return The flipped Y coordinate.
+     */
+    public abstract double flipY(double y);
+
+    /**
+     * Flips the heading.
+     *
+     * @param heading The heading to flip.
+     * @return The flipped heading.
+     */
+    public abstract double flipHeading(double heading);
+  }
+
+  private static record YearInfo(Flipper flipper, double fieldLength, double fieldWidth) {}
+
+  // TODO: Update and expand this map
+  private static final HashMap<Integer, YearInfo> flipperMap =
+      new HashMap<Integer, YearInfo>() {
+        {
+          put(2020, new YearInfo(Flipper.ROTATIONALLY_MIRRORED, 16.5811, 8.19912));
+          put(2021, new YearInfo(Flipper.ROTATIONALLY_MIRRORED, 16.5811, 8.19912));
+          put(2022, new YearInfo(Flipper.ROTATIONALLY_MIRRORED, 16.5811, 8.19912));
+          put(2023, new YearInfo(Flipper.VERTICALLY_MIRRORED, 16.5811, 8.19912));
+          put(2024, new YearInfo(Flipper.VERTICALLY_MIRRORED, 16.5811, 8.19912));
+        }
+      };
+
+  private static YearInfo activeYear = flipperMap.get(2024);
+
+  /**
+   * Get the flipper that is currently active for flipping coordinates. It's reccomended not to
+   * store this locally as the flipper may change.
+   *
+   * @return The active flipper.
+   */
+  public static Flipper getFlipper() {
+    return activeYear.flipper;
+  }
+
+  /**
+   * Returns if you are on red alliance, if the alliance is unknown it will return false.
+   *
+   * @return If you are on red alliance.
+   */
+  public static boolean onRed() {
+    return DriverStation.getAlliance().orElse(Alliance.Blue) == Alliance.Red;
```

Comment:
This should throw or something if there isn't an alliance color, not assume blue.

Reply:
I disagree, the driver station is often not available when robot code is first booting. This would result in many ds connectivity checks or the user bypassing these functions altogether. So it's really either they don't throw or we don't include these methods.

Reply:
> this would result in many ds connectivity checks or the user bypassing these functions altogether to prevent silly crashes.

No, that's exactly the point. Throwing and then the code restarting while the DS is disconnected is far better than the code starting up and latching to the wrong alliance. See https://github.com/wpilibsuite/allwpilib/pull/5229, https://github.com/wpilibsuite/allwpilib/pull/5230